hawkular-alerts-engine/src/main/resources/org/hawkular/alerts/engine/rules/ConditionMatch.drl

```diff
@@ -263,6 +266,41 @@ rule Rate
         insert( ce );
 end
 
+// NelsonCondition is a bit different (but somewhat like RateCondition) in that it looks for violations
+// of NelsonRules (https://en.wikipedia.org/wiki/Nelson_rules), as configured by the NelsonCondition.  To do this
+// we need to record a set of previous values for given dataId.  This rule creates the NelsonData for tracking
+// the given dataId, it does not perform any work on the data (see next rule).
+rule ProvideInitialNelsonData
+    when
+        $t : Trigger( $tenantId : tenantId, $tid : id, $tmode : mode, $tsource : source )
+        $c : NelsonCondition ( tenantId == $tenantId, triggerId == $tid, triggerMode == $tmode, $did : dataId )
+        $d : Data( tenantId == $tenantId, source == $tsource, id == $did )
+        not  NelsonData( data.tenantId == $tenantId, data.source == $tsource, data.id == $did )
+    then
+        NelsonData nelsonData = new NelsonData( $d );
+        if (log != null && log.isDebugEnabled()) {
+            log.debugf("ProvideInitialNelsonData: %s", nelsonData);
+        }
+        insert( nelsonData );
+end
+
+// Given current data and a NelsonData for a given dataId, perform a NelsonCondition evaluation.
+rule Nelson
+    when
+        $t  : Trigger( $tenantId : tenantId, $tid : id, $tmode : mode, $tsource : source )
+        $c  : NelsonCondition ( tenantId == $tenantId, triggerId == $tid, triggerMode == $tmode, $did : dataId )
+        $d  : Data( tenantId == $tenantId, source == $tsource, id == $did )
+        $nd : NelsonData( data.tenantId == $tenantId, data.source == $tsource, data.id == $did )

```

Comment:
Not sure if I have seen that in the code, but the NelsonData is removed when a condition is deleted ?
Just thinking on potential memory left on the engine.